core/trino-main/src/main/java/io/trino/execution/scheduler/SqlQueryScheduler.java

```diff
@@ -1548,6 +1591,327 @@ public void addStateChangeListener(StateChangeListener<DistributedStagesSchedule
+
+            Session session = queryStateMachine.getSession();
+            int hashPartitionCount = getHashPartitionCount(session);
+            Map<PartitioningHandle, BucketToPartition> bucketToPartitionCacheMap = new HashMap<>();
+            Function<PartitioningHandle, BucketToPartition> bucketToPartitionMapCache = partitioningHandle ->
+                    bucketToPartitionCacheMap.computeIfAbsent(partitioningHandle, handle -> createBucketToPartitionMap(session, hashPartitionCount, handle, nodePartitioningManager));
+
+            ImmutableList.Builder<FaultTolerantStageScheduler> schedulers = ImmutableList.builder();
+            Map<PlanFragmentId, Exchange> exchanges = new HashMap<>();
+
+            FixedCountNodeAllocator nodeAllocator = new FixedCountNodeAllocator(nodeScheduler, session, 1);
+            ScheduledFuture<?> nodeUpdateTask = scheduledExecutorService.scheduleAtFixedRate(nodeAllocator::updateNodes, 5, 5, SECONDS);
+
+            try {
+                // top to bottom order
+                List<SqlStageExecution> distributedStagesInTopologicalOrder = stages.getDistributedStagesInTopologicalOrder();
+                // bottom to top order
+                List<SqlStageExecution> distributedStagesInReverseTopologicalOrder = reverse(distributedStagesInTopologicalOrder);
+
+                ImmutableSet.Builder<PlanFragmentId> coordinatorConsumedFragmentsBuilder = ImmutableSet.builder();
+
+                for (SqlStageExecution stage : distributedStagesInReverseTopologicalOrder) {
+                    PlanFragment fragment = stage.getFragment();
+                    Optional<SqlStageExecution> parent = stages.getParent(stage.getStageId());
+                    TaskLifecycleListener taskLifecycleListener;
+                    Optional<Exchange> exchange;
+                    if (parent.isEmpty() || parent.get().getFragment().getPartitioning().isCoordinatorOnly()) {
+                        // output will be consumed by coordinator
+                        exchange = Optional.empty();
+                        taskLifecycleListener = coordinatorTaskLifecycleListener;
+                        coordinatorConsumedFragmentsBuilder.add(fragment.getId());
+                    }
+                    else {
+                        // create external exchange
+                        exchange = Optional.of(exchangeManager.create(new ExchangeContext(session.getQueryId(), stage.getStageId().getId()), hashPartitionCount));
+                        exchanges.put(fragment.getId(), exchange.get());
+                        taskLifecycleListener = TaskLifecycleListener.NO_OP;
+                    }
+
+                    ImmutableMap.Builder<PlanFragmentId, Exchange> inputExchanges = ImmutableMap.builder();
+                    for (SqlStageExecution child : stages.getChildren(fragment.getId())) {
+                        PlanFragmentId childFragmentId = child.getFragment().getId();
```

Comment:
name it `sourceFragmentId`

Reply:
Then I would probably need to rename `SqlStageExecution child`. I agree that the parent / child naming is a little weird, but it is the way we name the source stages in many places (e.g.: in `SubPlan`)